File: .pi/extensions/session-harness/orchestrate.js
#!/usr/bin/env node
/**
 * Phase 0.2 — short-session orchestrator with chain loop.
 *
 * Uses the pi SDK to create sessions in-process (no subprocess).
 * Reads specs/ files to drive a chain of stories, one session per story,
 * until the epic is complete.
 *
 * Usage:
 *     node orchestrate.js                          # chain all pending stories in active epic
 *     node orchestrate.js e05s01                   # run a single story
 *     node orchestrate.js --epic e05               # chain a specific epic
 *
 * Design:
 *     Each story gets a fresh in-process AgentSession with a clean
 *     context window. Sessions are ephemeral (in-memory). The chain
 *     advances based on execution-status.yaml's active.story_id cursor.
 */

import { readFileSync, writeFileSync, readdirSync, existsSync } from "node:fs";
import { join } from "node:path";
import { parse } from "yaml";
import {
  createAgentSession,
  DefaultResourceLoader,
  getAgentDir,
  SessionManager,
} from "@earendil-works/pi-coding-agent";

// --- Resolution -----------------------------------------------------------

// Walk up from the script's directory to find a project root
// that contains specs/state.yaml.
function resolveProjectDir(scriptDir) {
  const cwd = process.cwd();
  if (existsSync(join(cwd, "specs", "state.yaml"))) return cwd;
  let dir = scriptDir;
  for (let i = 0; i < 6; i++) {
    const candidate = join(dir, "..");
    if (existsSync(join(candidate, "specs", "state.yaml"))) return candidate;
    dir = candidate;
  }
  return cwd;
}

// --- Helpers ---------------------------------------------------------------

function loadYaml(path) {
  if (!existsSync(path)) {
    throw new Error(`File not found: ${path}`);
  }
  return parse(readFileSync(path, "utf8"));
}

function findEpicDir(projectDir, epicId) {
  const epicsDir = join(projectDir, "specs", "epics");
  if (!existsSync(epicsDir)) return null;
  for (const entry of readdirSync(epicsDir)) {
    if (entry.startsWith(epicId + "-")) {
      return join(epicsDir, entry);
    }
  }
  return null;
}

function findNextStory(epic, execStatus) {
  // The chain uses execution-status.yaml's `active.story_id` as a cursor.
  const currentId = execStatus.active?.story_id;
  if (!currentId) {
    return epic.stories?.[0] ?? null;
  }
  const idx = epic.stories.findIndex((s) => s.id === currentId);
  if (idx === -1) return epic.stories?.[0] ?? null;
  return epic.stories[idx + 1] ?? null;
}

function formatPrompt(story, epic, epicId, contextFiles) {
  const tasks = story.tasks ?? [];
  const taskLines = [];
  for (let i = 0; i < tasks.length; i++) {
    const t = tasks[i];
    const verify = t.verify ? `\n   Verify: \`${t.verify}\`` : "";
    taskLines.push(`${i + 1}. ${t.description}${verify}`);
  }

  // Build context block from pre-digested source files
  let contextBlock = "";
  if (contextFiles && contextFiles.length > 0) {
    const contextLines = [];
    for (const cf of contextFiles) {
      if (cf.content) {
        // Truncate large files: first 200 lines + last 50
        const lines = cf.content.split("\n");
        let shown;
        if (lines.length > 300) {
          shown = lines.slice(0, 200).join("\n")
            + `\n\n... (${lines.length - 250} lines omitted) ...\n\n`
            + lines.slice(-50).join("\n");
        } else {
          shown = cf.content;
        }
        contextLines.push(
          `### ${cf.path}\n\n\`\`\`\n${shown}\n\`\`\``
        );
      }
    }
    if (contextLines.length > 0) {
      contextBlock = [
        "\n## Context",
        "\nPre-digested source files for this story:\n",
        contextLines.join("\n\n"),
      ].join("\n");
    }
  }

  return [
    `# Story ${story.id}: ${story.title}`,
    "",
    `Epic: ${epicId} — ${epic.title}`,
    epic.description ? `\n${epic.description.trim()}\n` : "",
    contextBlock,
    "## Your job",
    `Execute story ${story.id}. Work in small, tested steps.`,
    "",
    "## Tasks",
    taskLines.length > 0 ? taskLines.join("\n") : "(no tasks defined for this story)",
    "",
    "## Constraints",
    "- Work on the active feature branch, NOT on master.",
    "- Run the verification command for each task before moving on.",
    "- When the story is complete, update specs/execution-status.yaml:",
    `  - Set active.epic_id to ${epicId}`,
    `  - Set active.story_id to ${story.id}`,
    "  - Set active.status to done",
    "  - Set active.completed_at to the current ISO timestamp.",
    "- Do NOT start another story in this session.",
    "",
    "Begin.",
  ].join("\n");
}

// --- Metrics tracker -------------------------------------------------------

class StoryMetrics {
  constructor(storyId) {
    this.storyId = storyId;
    this.modelProvider = null;
    this.modelId = null;
    this.startTime = null;
    this.endTime = null;
    this.turns = 0;
    this.toolCalls = {}; // tool name -> count
    this.toolErrors = {}; // tool name -> count
    this.tokens = { input: 0, output: 0, cacheRead: 0, cacheWrite: 0 };
    this.costUsd = 0;
    this.thinkingChars = 0;
    this.firstToolAt = null;
    this.lastToolAt = null;
    this.toolLatencyMs = 0;
  }

  attachSession(session) {
    if (session.model) {
      this.modelProvider = session.model.provider;
      this.modelId = session.model.id;
    }
  }

  record(event) {
    switch (event.type) {
      case "agent_start":
        if (!this.startTime) this.startTime = Date.now();
        this.turns++;
        break;
      case "agent_end":
        // each turn completed; not a final signal
        break;
      case "tool_execution_start": {
        const name = event.toolName;
        this.toolCalls[name] = (this.toolCalls[name] || 0) + 1;
        const now = Date.now();
        if (!this.firstToolAt) this.firstToolAt = now;
        this.lastToolAt = now;
        break;
      }
      case "tool_execution_end": {
        if (event.isError) {
          const name = event.toolName;
          this.toolErrors[name] = (this.toolErrors[name] || 0) + 1;
        }
        if (this.lastToolAt) this.toolLatencyMs += Date.now() - this.lastToolAt;
        break;
      }
      case "message_update": {
        const e = event.assistantMessageEvent;
        if (e?.type === "thinking_delta" && e.delta) {
          this.thinkingChars += e.delta.length;
        }
        break;
      }
      case "message_end": {
        const msg = event.message;
        if (msg?.usage) {
          this.tokens.input += msg.usage.input || 0;
          this.tokens.output += msg.usage.output || 0;
          this.tokens.cacheRead += msg.usage.cacheRead || 0;
          this.tokens.cacheWrite += msg.usage.cacheWrite || 0;
          if (msg.usage.cost?.total) this.costUsd += msg.usage.cost.total;
        }
        // Capture model from the first assistant message
        if (msg?.model && !this.modelId) {
          this.modelId = msg.model;
          this.modelProvider = msg.provider;
        }
        break;
      }
    }
  }

  finish() {
    this.endTime = Date.now();
  }

  durationSec() {
    if (!this.startTime || !this.endTime) return 0;
    return ((this.endTime - this.startTime) / 1000).toFixed(1);
  }

  print() {
    const totalTools = Object.values(this.toolCalls).reduce((a, b) => a + b, 0);
    const toolList = Object.entries(this.toolCalls)
      .map(([name, n]) => `${name}=${n}`)
      .join(" ");
    const errorList = Object.entries(this.toolErrors)
      .map(([name, n]) => `${name}=${n}`)
      .join(" ");
    const model = this.modelProvider && this.modelId
      ? `${this.modelProvider}/${this.modelId}`
      : "(unknown)";
    const tokens = [
      `in=${formatNum(this.tokens.input)}`,
      `out=${formatNum(this.tokens.output)}`,
      `cacheR=${formatNum(this.tokens.cacheRead)}`,
      `cacheW=${formatNum(this.tokens.cacheWrite)}`,
    ].join(" ");

    const lines = [
      `=== metrics: ${this.storyId} ===`,
      `  model:    ${model}`,
      `  duration: ${this.durationSec()}s`,
      `  turns:    ${this.turns}`,
      `  tools:    ${totalTools} total (${toolList || "none"})`,
      `  errors:   ${errorList || "none"}`,
      `  tokens:   ${tokens}`,
      `  cost:     $${this.costUsd.toFixed(4)}`,
      `  thinking: ${formatNum(this.thinkingChars)} chars`,
    ];
    process.stdout.write(`\n${lines.join("\n")}\n`);
  }

  toJSON() {
    return {
      storyId: this.storyId,
      model: this.modelProvider && this.modelId
        ? `${this.modelProvider}/${this.modelId}`
        : null,
      durationSec: parseFloat(this.durationSec()),
      turns: this.turns,
      toolCalls: this.toolCalls,
      toolErrors: this.toolErrors,
      tokens: this.tokens,
      costUsd: this.costUsd,
      thinkingChars: this.thinkingChars,
    };
  }
}

function formatNum(n) {
  if (n >= 1000) return (n / 1000).toFixed(1) + "K";
  return String(n);
}

// --- Session runner --------------------------------------------------------

async function runStory(projectDir, epicId, epic, story, opts = {}) {
  const promptDir = opts.epicDir || join(projectDir, "specs", "epics", epicId);
  // Load context files if the story defines them
  const contextFiles = [];
  if (story.context && Array.isArray(story.context)) {
    for (const relPath of story.context) {
      const absPath = join(projectDir, relPath);
      if (existsSync(absPath)) {
        try {
          const content = readFileSync(absPath, "utf8");
          contextFiles.push({ path: relPath, content });
        } catch {}
      }
    }
  }
  const prompt = formatPrompt(story, epic, epicId, contextFiles);
  const sessionName = `story-${story.id}`;
  const metrics = new StoryMetrics(story.id);
  const verbose = opts.verbose || false;

  // Write prompt to a file for inspection
  try {
    writeFileSync(join(promptDir, "last-prompt.md"), prompt, "utf8");
  } catch {}

  // Create a fresh in-memory session
  const loader = new DefaultResourceLoader({
    cwd: projectDir,
    agentDir: getAgentDir(),
  });
  await loader.reload();

  const { session } = await createAgentSession({
    cwd: projectDir,
    sessionManager: SessionManager.inMemory(projectDir),
    resourceLoader: loader,
    tools: ["read", "bash", "write", "edit", "grep", "find", "ls"],
  });

  metrics.attachSession(session);

  let settled = false;
  const unsub = session.subscribe((event) => {
    metrics.record(event);
    if (verbose) return; // verbose mode prints raw events instead
    switch (event.type) {
      case "message_update": {
        const e = event.assistantMessageEvent;
        if (e?.type === "text_delta") process.stdout.write(e.delta);
        else if (e?.type === "text_end") process.stdout.write("\n");
        else if (e?.type === "thinking_delta" && opts.showThinking) {
          // optional: show thinking in dim text
          process.stdout.write(`\x1b[2m[thinking] ${e.delta}\x1b[0m`);
        }
        break;
      }
      case "tool_execution_start":
        process.stdout.write(`\n[${sessionName}] (tool: ${event.toolName})\n`);
        break;
      case "tool_execution_end":
        if (event.isError) process.stdout.write(`[${sessionName}] (tool errored)\n`);
        break;
      case "agent_settled":
        settled = true;
        break;
    }
  });

  process.stdout.write(`\n=== ${sessionName} ===\n`);
  process.stdout.write(`${new Date().toISOString()} Starting ${story.id}\n\n`);

  try {
    await session.prompt(prompt);
  } catch (err) {
    process.stderr.write(`[${sessionName}] prompt error: ${err.message}\n`);
    metrics.finish();
    unsub();
    session.dispose();
    metrics.print();
    return "error";
  }

  if (!settled) await new Promise((r) => setTimeout(r, 1000));

  unsub();
  session.dispose();
  metrics.finish();

  // Result: did the agent mark the story done?
  const execPath = join(projectDir, "specs", "execution-status.yaml");
  if (!existsSync(execPath)) {
    metrics.print();
    return "error";
  }
  const status = loadYaml(execPath);
  const result = (status.active?.story_id === story.id && status.active?.status === "done")
    ? "done"
    : "pending";
  metrics.print();
  return result;
}

// --- Main ------------------------------------------------------------------

async function main() {
  const args = process.argv.slice(2);
  const scriptDir = new URL(".", import.meta.url).pathname;
  const projectDir = resolveProjectDir(scriptDir);

  let explicitEpic = null;
  let explicitStory = null;
  let opts = { verbose: false, showThinking: false };
  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--epic" && i + 1 < args.length) explicitEpic = args[++i];
    else if (args[i] === "--verbose") opts.verbose = true;
    else if (args[i] === "--show-thinking") opts.showThinking = true;
    else if (!args[i].startsWith("--")) explicitStory = args[i];
  }

  if (opts.verbose) {
    opts.showThinking = true; // verbose implies show thinking
  }

  // Read state
  const statePath = join(projectDir, "specs", "state.yaml");
  const state = loadYaml(statePath);
  const epicId = explicitEpic ?? state.active_epic_id;
  if (!epicId) {
    console.error(`FAIL: no active_epic_id — set it in specs/state.yaml or pass --epic`);
    process.exit(1);
  }

  const epicDir = findEpicDir(projectDir, epicId);
  if (!epicDir) {
    console.error(`FAIL: epic directory not found for ${epicId}`);
    process.exit(1);
  }
  const epicPath = join(epicDir, "epic.yaml");
  const epic = loadYaml(epicPath);

  const execPath = join(projectDir, "specs", "execution-status.yaml");
  let execStatus = loadYaml(execPath);

  // Single-story mode
  if (explicitStory) {
    const target = epic.stories.find((s) => s.id === explicitStory);
    if (!target) {
      console.error(`FAIL: story ${explicitStory} not found in epic ${epicId}`);
      process.exit(1);
    }
    console.log(`\n--- Single: ${explicitStory} ---\n`);
    const result = await runStory(projectDir, epicId, epic, target, { ...opts, epicDir });
    console.log(`\nResult: ${result}`);
    process.exit(result === "done" ? 0 : 1);
  }

  // Chain mode
  console.log(`\n--- Chain: epic ${epicId} ---`);
  let count = 0;
  const start = Date.now();
  let next = findNextStory(epic, execStatus);
  let lastId = null;

  while (next) {
    if (next.id === lastId) {
      console.log(`[chain] Cursor stuck at ${next.id}. Stopping.`);
      break;
    }
    lastId = next.id;
    count++;
    const result = await runStory(projectDir, epicId, epic, next, { ...opts, epicDir });
    console.log(`\n[chain] ${next.id} → ${result}`);
    if (result !== "done") {
      console.log(`[chain] Chain stopped — ${next.id} not marked done.`);
      break;
    }
    execStatus = loadYaml(execPath);
    next = findNextStory(epic, execStatus);
  }

  const elapsed = ((Date.now() - start) / 1000).toFixed(1);
  console.log(`\n[chain] Epic ${epicId}: ran ${count} story/stories in ${elapsed}s`);
  process.exit(0);
}

main().catch((err) => {
  console.error("FATAL:", err);
  process.exit(1);
});

File: .opencode/extensions/session-harness/orchestrate.py
#!/usr/bin/env python3
"""
Opencode session harness — chain loop for epic execution.

Reads specs/ to drive a chain of stories, one subagent per story,
until the epic is complete.

Usage:
    python orchestrate.py                    # chain all pending stories in active epic
    python orchestrate.py e08s01             # run a single story
    python orchestrate.py --epic e08         # chain a specific epic
"""

from __future__ import annotations

import argparse
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import yaml

# --- Constants -------------------------------------------------------------

_CONTEXT_TRUNCATE_LINE_LIMIT = 300
_CONTEXT_HEAD_LINES = 200
_CONTEXT_TAIL_LINES = 50


# --- YAML helpers -----------------------------------------------------------

def load_yaml(path: Path) -> dict[str, Any]:
    """Load a YAML file, returning an empty dict if it doesn't exist."""
    if not path.exists():
        return {}
    with open(path) as f:
        return yaml.safe_load(f)


def write_yaml(path: Path, data: dict[str, Any]) -> None:
    """Write a YAML file."""
    path.parent.mkdir(parents=True, exist_ok=True)
    with open(path, "w") as f:
        yaml.safe_dump(data, f, default_flow_style=False)


def load_epic(project_dir: Path, epic_id: str) -> dict[str, Any]:
    """Load an epic.yaml by epic ID (prefix match on directory name)."""
    epic_dir = find_epic_dir(project_dir, epic_id)
    epic_path = epic_dir / "epic.yaml"
    if not epic_path.exists():
        print(f"FAIL: epic.yaml not found at {epic_path}", file=sys.stderr)
        sys.exit(1)
    return load_yaml(epic_path)


def find_epic_dir(project_dir: Path, epic_id: str) -> Path:
    """Find the epic directory matching an epic ID prefix."""
    epics_dir = project_dir / "specs" / "epics"
    if not epics_dir.exists():
        print(f"FAIL: specs/epics/ not found", file=sys.stderr)
        sys.exit(1)
    for entry in sorted(epics_dir.iterdir()):
        if entry.is_dir() and entry.name.startswith(f"{epic_id}-"):
            return entry
    print(f"FAIL: epic directory not found for {epic_id}", file=sys.stderr)
    sys.exit(1)


# --- Cursor logic ----------------------------------------------------------

def find_next_story(
    epic: dict[str, Any], exec_status: dict[str, Any]
) -> dict[str, Any] | None:
    """Find the next uncompleted story using execution-status.yaml as cursor.

    The cursor is ``exec_status.active.story_id`` (last completed story).
    Returns the story after the cursor, or the first story if no cursor.
    Returns None if the epic has no more stories.
    """
    stories = epic.get("stories", [])
    if not stories:
        return None

    current_id = exec_status.get("active", {}).get("story_id")
    if not current_id:
        return stories[0]

    idx = next(
        (i for i, s in enumerate(stories) if s["id"] == current_id),
        -1,
    )
    if idx == -1:
        return stories[0]
    if idx + 1 >= len(stories):
        return None
    return stories[idx + 1]


def load_cursor(project_dir: Path) -> dict[str, Any]:
    """Load execution-status.yaml."""
    return load_yaml(project_dir / "specs" / "execution-status.yaml")


def update_cursor(
    project_dir: Path, epic_id: str, story_id: str, status: str = "done"
) -> None:
    """Mark a story complete in execution-status.yaml."""
    exec_path = project_dir / "specs" / "execution-status.yaml"
    data = load_yaml(exec_path)
    data["active"] = {
        "epic_id": epic_id,
        "story_id": story_id,
        "status": status,
        "completed_at": datetime.now(timezone.utc).isoformat(),
    }
    write_yaml(exec_path, data)


def load_active_epic(project_dir: Path) -> str:
    """Load the active epic ID from state.yaml."""
    state = load_yaml(project_dir / "specs" / "state.yaml")
    epic_id = state.get("active_epic_id")
    if not epic_id:
        print("FAIL: no active_epic_id in specs/state.yaml", file=sys.stderr)
        sys.exit(1)
    return epic_id


# --- Prompt formatting -----------------------------------------------------

def load_context_files(
    project_dir: Path, story: dict[str, Any]
) -> list[dict[str, str]]:
    """Pre-digest context files specified in the story definition."""
    context_refs = story.get("context", [])
    if not context_refs:
        return []

    context_files: list[dict[str, str]] = []
    for rel_path in context_refs:
        abs_path = project_dir / rel_path
        try:
            if not abs_path.exists():
                continue
            content = abs_path.read_text()
            lines = content.split("\n")
            if len(lines) > _CONTEXT_TRUNCATE_LINE_LIMIT:
                truncated = (
                    "\n".join(lines[:_CONTEXT_HEAD_LINES])
                    + f"\n\n... ({len(lines) - _CONTEXT_TAIL_LINES} lines omitted) ...\n\n"
                    + "\n".join(lines[-_CONTEXT_TAIL_LINES:])
                )
            else:
                truncated = content
            context_files.append({"path": str(abs_path), "content": truncated})
        except (OSError, PermissionError) as exc:
            print(f"[harness] Warning: could not read context file {abs_path}: {exc}")
    return context_files


def format_prompt(
    story: dict[str, Any],
    epic: dict[str, Any],
    epic_id: str,
    context_files: list[dict[str, str]] | None = None,
) -> str:
    """Build the story prompt from epic/story data and optional context."""
    task_lines: list[str] = []
    tasks = story.get("tasks", [])
    for i, t in enumerate(tasks, 1):
        line = f"{i}. {t['description']}"
        verify = t.get("verify", "")
        if verify:
            line += f"\n   Verify: `{verify}`"
        task_lines.append(line)

    # Context block
    context_block = ""
    if context_files:
        sections: list[str] = []
        for cf in context_files:
            sections.append(f"### {cf['path']}\n\n```\n{cf['content']}\n```")
        if sections:
            context_block = "\n## Context\n\n" + "\n\n".join(sections)

    epic_desc = epic.get("description", "")

    return f"""# Story {story['id']}: {story['title']}

Epic: {epic_id} — {epic['title']}
{epic_desc.strip()}
{context_block}
## Your job
Execute story {story['id']}. Work in small, tested steps.

## Tasks
{chr(10).join(task_lines) or '(no tasks defined)'}

## Constraints
- Work on the active feature branch, NOT on master.
- Run the verification command for each task before moving on.
- When the story is complete, update specs/execution-status.yaml:
  - Set active.epic_id to {epic_id}
  - Set active.story_id to {story['id']}
  - Set active.status to done
  - Set active.completed_at to the current ISO timestamp.
- Commit all changes to the worktree before marking the story done.
  Use a conventional commit message: `{story['id']}: <short description>`.
  Include `last-prompt.md` in the commit — it records the original intent.
- Do NOT start another story in this session.

Begin."""


# --- Story runner ----------------------------------------------------------

def run_story(
    project_dir: Path,
    epic_id: str,
    epic: dict[str, Any],
    story: dict[str, Any],
    *,
    dry_run: bool = False,
) -> str:
    """Prepare a story for execution and print the task parameters.

    In dry_run mode, writes the prompt to last-prompt.md and prints it.
    For actual execution, the orchestrator outputs the task() call that
    the opencode agent should invoke.

    Returns the final status read from execution-status.yaml after the
    agent completes: "done" or "pending".
    """
    # Find the epic directory for artifact output
    epic_dir = find_epic_dir(project_dir, epic_id)

    # Load context files
    context_files = load_context_files(project_dir, story)

    # Format prompt
    prompt = format_prompt(story, epic, epic_id, context_files)

    # Write prompt for inspection
    prompt_path = epic_dir / "last-prompt.md"
    prompt_path.write_text(prompt)
    print(f"[harness] Prompt written to {prompt_path}")

    # Commit the prompt so we record what the author intended
    try:
        subprocess.run(
            ["git", "add", str(prompt_path)],
            cwd=project_dir,
            check=True,
            capture_output=True,
        )
        subprocess.run(
            ["git", "commit", "-m", f"chore(harness): dispatch {story['id']}"],
            cwd=project_dir,
            check=True,
            capture_output=True,
        )
        print(f"[harness] Committed last-prompt.md")
    except subprocess.CalledProcessError as exc:
        print(f"[harness] Warning: could not commit last-prompt.md: {exc}")

    if dry_run:
        print("\n--- PROMPT ---")
        print(prompt)
        print("--- END PROMPT ---")
        return "dry_run"

    # Print the execution instructions for the opencode agent
    print(f"\n[harness] Story {story['id']}: '{story['title']}'")
    print(f"[harness] To execute this story in opencode, call:")
    print(f"[harness]")
    print(f"[harness]   task(")
    print(f"[harness]       category='deep',")
    print(f"[harness]       load_skills=['test-driven-development', 'verification-before-completion'],")
    print(f"[harness]       prompt=read_file('{prompt_path}')")
    print(f"[harness]   )")
    print(f"[harness]")
    print(f"[harness] After execution, check specs/execution-status.yaml for completion.")

    # Read the cursor to get final status
    exec_status = load_cursor(project_dir)
    active = exec_status.get("active", {})
    if active.get("story_id") == story["id"] and active.get("status") == "done":
        return "done"
    return "pending"


# --- Chain runner ----------------------------------------------------------

def chain(
    project_dir: Path,
    epic_id: str,
    *,
    dry_run: bool = False,
) -> None:
    """Chain through all pending stories in an epic."""
    epic = load_epic(project_dir, epic_id)
    exec_status = load_cursor(project_dir)
    count = 0
    last_id: str | None = None

    print(f"\n--- Chain: epic {epic_id} ---\n")

    next_story = find_next_story(epic, exec_status)
    while next_story:
        if next_story["id"] == last_id:
            print(f"[chain] Cursor stuck at {next_story['id']}. Stopping.")
            break

        last_id = next_story["id"]
        count += 1

        result = run_story(project_dir, epic_id, epic, next_story, dry_run=dry_run)
        print(f"\n[chain] {next_story['id']} → {result}")

        if result != "done":
            print(f"[chain] Chain stopped — {next_story['id']} not marked done.")
            break

        # Reload cursor for next iteration
        exec_status = load_cursor(project_dir)
        next_story = find_next_story(epic, exec_status)

    print(f"\n[chain] Epic {epic_id}: ran {count} story/stories.")


# --- CLI entry point -------------------------------------------------------

def resolve_project_dir() -> Path:
    """Find the project root containing specs/state.yaml."""
    cwd = Path.cwd()
    if (cwd / "specs" / "state.yaml").exists():
        return cwd

    script_dir = Path(__file__).resolve().parent
    for _ in range(6):
        if (script_dir / "specs" / "state.yaml").exists():
            return script_dir.resolve()
        script_dir = script_dir.parent

    print("FAIL: could not find project root (no specs/state.yaml)", file=sys.stderr)
    sys.exit(1)


def main() -> None:
    parser = argparse.ArgumentParser(description="Opencode session harness")
    parser.add_argument(
        "story_id", nargs="?", help="Run a single story (e.g. e08s01)"
    )
    parser.add_argument(
        "--epic", help="Chain a specific epic (default: from specs/state.yaml)"
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Print the prompt without executing",
    )
    parser.add_argument(
        "--model",
        help="Model to use for story execution",
    )
    args = parser.parse_args()

    project_dir = resolve_project_dir()

    # Resolve epic ID
    epic_id = args.epic or load_active_epic(project_dir)

    if args.story_id:
        # Single-story mode
        epic = load_epic(project_dir, epic_id)
        stories = epic.get("stories", [])
        target = next((s for s in stories if s["id"] == args.story_id), None)
        if not target:
            print(f"FAIL: story {args.story_id} not found in epic {epic_id}")
            sys.exit(1)

        result = run_story(project_dir, epic_id, epic, target, dry_run=args.dry_run)
        print(f"\nResult: {result}")
        sys.exit(0 if result == "done" else 1)
    else:
        # Chain mode
        chain(project_dir, epic_id, dry_run=args.dry_run)


if __name__ == "__main__":
    main()